File: 15.三数之和.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=15 lang=java
 *
 * [15] 三数之和
 * 找三个和为 0 的数
 */

// @lc code=start
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            int target = 0 - nums[i];
            List<List<Integer>> twoSumAns = twoSum(nums, i + 1, target);
            for (List<Integer> twoSum : twoSumAns) {
                ans.add(Arrays.asList(nums[i], twoSum.get(0), twoSum.get(1)));
            }
            while (i < nums.length - 1 && nums[i] == nums[i + 1]) {
                ++i;
            }
        }
        return ans;
    }

    // 要保证传入的 nums 必须是有序的。
    // 返回的是两个值，而不是下标。
    private List<List<Integer>> twoSum(int[] nums, int start, int target) {
        List<List<Integer>> res = new ArrayList<>();

        int l = start, r = nums.length - 1;
        // 在 while 中做索引移动，一定要重复判断索引关系。
        while (l < r) {
            int lv = nums[l], rv = nums[r];
            if (lv + rv == target) {
                res.add(Arrays.asList(lv, rv));
                while (l < r && nums[l] == lv) {
                    l++;
                }
                while (l < r && nums[r] == rv) {
                    r--;
                }
            } else if (lv + rv < target) {
                while (l < r && nums[l] == lv) {
                    l++;
                }
            } else {
                while (l < r && nums[r] == rv) {
                    r--;
                }
            }
        }

        return res;
    }
}
// @lc code=end

File: 15.三数之和.go
import "sort"

/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start
// 思路：双指针。排序之后遍历 nums，对于其中的每个元素，再用类似两数之和的思路去找有没有和是 -index 的。
func threeSum(nums []int) [][]int {
	sort.Ints(nums)

	var (
		res = make([][]int, 0)
		n   = len(nums)
	)

	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		if nums[i]+nums[i+1]+nums[i+2] > 0 {
			break
		}
		if nums[i]+nums[n-2]+nums[n-1] < 0 {
			continue
		}

		left, right := i+1, n-1
		for left < right {
			s := nums[i] + nums[left] + nums[right]
			if s > 0 {
				right--
			} else if s < 0 {
				left++
			} else {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				left++
				for left < right && nums[left] == nums[left-1] {
					left++
				}
				right--
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			}
		}
	}

	return res
}

// @lc code=end

